File: src/main/java/net/illuc/kontraption/KontraptionBlocks.java
package net.illuc.kontraption;

import mekanism.common.block.prefab.BlockBasicMultiblock;
import mekanism.common.registration.impl.BlockDeferredRegister;
import mekanism.common.registration.impl.BlockRegistryObject;
import net.illuc.kontraption.blocks.*;
import net.minecraft.network.chat.Component;
import net.minecraft.world.item.BlockItem;
import net.minecraft.world.level.block.Block;
import net.minecraft.world.level.block.state.BlockBehaviour;
import net.minecraft.world.level.material.MapColor;

import java.util.List;

public class KontraptionBlocks {
    private KontraptionBlocks() {
    }

    public record PlushieData(
            String name,
            String role,
            String tooltip
    ) {
    }

    ;


    public static final BlockDeferredRegister BLOCKS = new BlockDeferredRegister(Kontraption.MODID);

    // <-----BLOCKS----->
    //public static final BlockRegistryObject<BlockTile.BlockTileModel<TileEntityIonThruster, Machine<TileEntityIonThruster>>, ItemBlockMachine> ION_THRUSTER = BLOCKS.register("ion_thruster", () -> new BlockTile.BlockTileModel<>(KontraptionBlockTypes.ION_THRUSTER), ItemBlockMachine::new);
    public static final BlockRegistryObject<BlockIonThruster, BlockItem> ION_THRUSTER = BLOCKS.register("ion_thruster", () -> new BlockIonThruster(KontraptionBlockTypes.ION_THRUSTER));
    public static final BlockRegistryObject<ShipControlInterfaceBlock, BlockItem> SHIP_CONTROL_INTERFACE = BLOCKS.register("ship_control_interface", () -> new ShipControlInterfaceBlock(KontraptionBlockTypes.SHIP_CONTROL_INTERFACE));
    public static final BlockRegistryObject<BlockGyro, BlockItem> GYRO = BLOCKS.register("gyro", () -> new BlockGyro(KontraptionBlockTypes.GYRO));
    //needs work
    public static final BlockRegistryObject<BlockWheel, BlockItem> WHEEL = BLOCKS.register("wheel", () -> new BlockWheel(KontraptionBlockTypes.WHEEL));
    public static final BlockRegistryObject<BlockBasicMultiblock, BlockItem> LIQUID_FUEL_THRUSTER_CASING = BLOCKS.register("liquid_fuel_thruster_casing", () -> new BlockBasicMultiblock<>(KontraptionBlockTypes.LIQUID_FUEL_THRUSTER_CASING, properties -> properties.mapColor(MapColor.COLOR_GRAY)));
    public static final BlockRegistryObject<BlockBasicMultiblock, BlockItem> LIQUID_FUEL_THRUSTER_VALVE = BLOCKS.register("liquid_fuel_thruster_valve", () -> new BlockBasicMultiblock<>(KontraptionBlockTypes.LIQUID_FUEL_THRUSTER_VALVE, properties -> properties.mapColor(MapColor.COLOR_GRAY)));
    public static final BlockRegistryObject<BlockBasicMultiblock, BlockItem> LIQUID_FUEL_THRUSTER_EXHAUST = BLOCKS.register("liquid_fuel_thruster_exhaust", () -> new BlockBasicMultiblock<>(KontraptionBlockTypes.LIQUID_FUEL_THRUSTER_EXHAUST, properties -> properties.mapColor(MapColor.COLOR_GRAY)));

    public static final BlockRegistryObject<Block, BlockItem> RUBBER_BLOCK = BLOCKS.register("rubber_block", () -> new Block(BlockBehaviour.Properties.of()));

    public static final BlockRegistryObject<BlockCannon, BlockItem> CANNON = BLOCKS.register("cannon", () -> new BlockCannon(KontraptionBlockTypes.CANNON));
    public static final BlockRegistryObject<BlockServo, BlockItem> SERVO = BLOCKS.register("servo", () -> new BlockServo(KontraptionBlockTypes.SERVO));
}

File: src/main/kotlin/net/illuc/kontraption/KontraptionBlocks.kt
package net.illuc.kontraption

import mekanism.common.block.interfaces.IHasDescription
import mekanism.common.block.prefab.BlockTile.BlockTileModel
import mekanism.common.content.blocktype.BlockTypeTile
import mekanism.common.content.blocktype.Machine
import mekanism.common.item.block.ItemBlockTooltip
import mekanism.common.item.block.machine.ItemBlockMachine
import mekanism.common.registration.impl.BlockDeferredRegister
import mekanism.common.registration.impl.BlockRegistryObject
import mekanism.common.tile.base.TileEntityMekanism
import net.illuc.kontraption.blockEntities.TileIonThruster
import net.minecraft.world.item.Item
import net.minecraft.world.level.block.Block
import java.util.function.Supplier

/*
public object KontraptionBlocks {
    private fun KontraptionBlocks() {}

    val BLOCKS = BlockDeferredRegister(Kontraption.MODID)


    //val TURBINE_CASING: BlockRegistryObject<BlockBasicMultiblock<TileIonThruster>, ItemBlockTooltip<BlockBasicMultiblock<TileIonThruster>>> = registerTooltipBlock("turbine_casing") { BlockBasicMultiblock<TILE>(KontraptionBlocks.ION_THRUSTER) }

    //val ION_THRUSTER = BLOCKS.register("ion_thruster", { BlockTileModel(KontraptionBlockTypes.ION_THRUSTER) }) { block: BlockTileModel<TileIonThruster?, Machine<TileIonThruster?>?>? -> ItemBlockMachine(block) }

    //val ION_THRUSTER: BlockRegistryObject<BlockTile<TileIonThruster?, Machine<TileIonThruster?>?>, ItemBlockMachine> = BLOCKS.register<BlockTile<TileIonThruster?, Machine<TileIonThruster?>?>, ItemBlockMachine>("ion_thruster", Supplier<BlockTile<TileIonThruster?, Machine<TileIonThruster?>?>> { BlockTile<TileEntityMekanism, BlockTypeTile<TileEntityMekanism>>(KontraptionBlockTypes., UnaryOperator<BlockBehaviour.Properties> { properties: BlockBehaviour.Properties -> properties.mapColor(BlockResourceInfo.STEEL.getMapColor()) }) }, Function<BlockTile<TileEntityDimensionalStabilizer?, Machine<TileEntityDimensionalStabilizer?>?>, ItemBlockMachine> { block: BlockTile<TileEntityDimensionalStabilizer?, Machine<TileEntityDimensionalStabilizer?>?>? -> ItemBlockMachine(block) })
    //val ION_THRUSTER: BlockRegistryObject<BlockTileModel<TileIonThruster?, Machine<TileIonThruster?>>, ItemBlockMachine>? = BLOCKS.register<BlockTileModel<TileIonThruster?, Machine<TileIonThruster?>>, ItemBlockMachine>("ion_thruster", { BlockTileModel<TileEntityMekanism?, BlockTypeTile<TileEntityMekanism>>(KontraptionBlockTypes.ION_THRUSTER) }) { block: BlockTileModel<TileIonThruster?, Machine<TileIonThruster?>>? -> ItemBlockMachine(block) }

    private fun <BLOCK> registerTooltipBlock(name: String, blockCreator: Supplier<BLOCK>): BlockRegistryObject<BLOCK, ItemBlockTooltip<BLOCK>>? where BLOCK : Block?, BLOCK : IHasDescription? {
        return BLOCKS.registerDefaultProperties(name, blockCreator) { block: BLOCK, properties: Item.Properties? -> ItemBlockTooltip(block, properties) }
    }

}*/